File: dashboard-web-react/src/features/cameras/cameraCone.test.ts
import { describe, expect, it } from 'vitest';
import { cameraConePolygon } from './cameraCone';

const CENTER = { lat: 12.68, lng: 108.05 };

describe('cameraConePolygon', () => {
  it('starts and ends at the camera position, framing a closed fan shape', () => {
    const points = cameraConePolygon(CENTER, 90, 60);
    expect(points[0]).toEqual([CENTER.lat, CENTER.lng]);
    expect(points[points.length - 1]).toEqual([CENTER.lat, CENTER.lng]);
  });

  it('returns one arc point per segment plus the two center points', () => {
    const points = cameraConePolygon(CENTER, 0, 90, 60, 8);
    expect(points).toHaveLength(8 + 1 + 2);
  });

  it('a camera facing due east (90°) moves the arc points east, not north/south', () => {
    const points = cameraConePolygon(CENTER, 90, 10, 60, 4);
    const arcPoints = points.slice(1, -1);
    for (const [lat, lng] of arcPoints) {
      expect(lng).toBeGreaterThan(CENTER.lng);
      expect(lat).toBeCloseTo(CENTER.lat, 3);
    }
  });

  it('a camera facing due north (0°) moves the arc points north, not east/west', () => {
    const points = cameraConePolygon(CENTER, 0, 10, 60, 4);
    const arcPoints = points.slice(1, -1);
    for (const [lat, lng] of arcPoints) {
      expect(lat).toBeGreaterThan(CENTER.lat);
      expect(lng).toBeCloseTo(CENTER.lng, 3);
    }
  });
});

File: mobile-app-officer/test/camera_cone_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:latlong2/latlong.dart' as latlong;
import 'package:bao_tin_officer/features/cameras/camera_cone.dart';

const _center = latlong.LatLng(12.68, 108.05);

void main() {
  test('starts and ends at the camera position, framing a closed fan shape', () {
    final points = cameraConePolygon(_center, 90, 60);
    expect(points.first, _center);
    expect(points.last, _center);
  });

  test('returns one arc point per segment plus the two center points', () {
    final points = cameraConePolygon(_center, 0, 90, segments: 8);
    expect(points.length, 8 + 1 + 2);
  });

  test('a camera facing due east (90°) moves the arc points east, not north/south', () {
    final points = cameraConePolygon(_center, 90, 10, segments: 4);
    final arcPoints = points.sublist(1, points.length - 1);
    for (final p in arcPoints) {
      expect(p.longitude, greaterThan(_center.longitude));
      expect(p.latitude, closeTo(_center.latitude, 0.001));
    }
  });

  test('a camera facing due north (0°) moves the arc points north, not east/west', () {
    final points = cameraConePolygon(_center, 0, 10, segments: 4);
    final arcPoints = points.sublist(1, points.length - 1);
    for (final p in arcPoints) {
      expect(p.latitude, greaterThan(_center.latitude));
      expect(p.longitude, closeTo(_center.longitude, 0.001));
    }
  });
}
